### ISSUE TITLE
encoder

### ISSUE THREAD

**Issue by jh12521** [takuseno/d3rlpy#414]
How to create own encoder, load pre training parameters, and ensure that the encoder does not participate in backpropagation during policy training?

**Comment by takuseno** [comment]
@jh12521 Sorry for the late response. This is a tutorial about the custom neural network model.
https://d3rlpy.readthedocs.io/en/v2.5.0/tutorials/customize_neural_network.html

You can easily freeze partial parameters by using `detach()`. Since that part is standard PyTorch usage, it's not a scope of d3rlpy. Let me close this issue.

**Comment by peymvn** [comment]
I had a similar question, and since I didn't find it in the documentation, I wanted to share it for future reference. The custom encoder is well-documented, however, if you've trained your base model using d3rlpy and want to freeze the encoder for finetuning, you can do so directly on the underlying PyTorch model without the need to create a custom encoder class. Here's how I did it for CQL:

`
encoder_factory = d3rlpy.models.VectorEncoderFactory(
    hidden_units=hidden_units,
    activation=activation_func
)

new_cql = d3rlpy.algos.DiscreteCQLConfig(
    alpha=conservative_alpha, 
    encoder_factory=encoder_factory, 
    learning_rate=learning_rate
).create(device='cuda')

new_cql.build_with_dataset(finetune_dataset)
new_cql.copy_q_function_from(base_model)

# Freeze the encoder (accessible via the "_impl" attribute)
for name, param in new_cql._impl._checkpointer.modules['q_funcs'].named_parameters():
    if 'encoder' in name:
        param.requires_grad = False

for name, param in new_cql._impl._checkpointer.modules['targ_q_funcs'].named_parameters():
    if 'encoder' in name:
        param.requires_grad = False

# Fine-tune the RL model
new_cql.fit(
    finetune_dataset,
    n_steps=n_steps
)
`

@takuseno, please correct me if I'm wrong.
